sessionctx/sessionstates/session_token.go

```diff
@@ -0,0 +1,338 @@
+	// - server B receives the token at 00:10:02, so the old cert should be valid for more than 10m after replacement.
+	oldCertValidTime = 15 * time.Minute
+)
+
+// SessionToken represents the token used to authenticate with the new server.
+type SessionToken struct {
+	Username   string    `json:"username"`
+	SignTime   time.Time `json:"sign-time"`
+	ExpireTime time.Time `json:"expire-time"`
+	Signature  []byte    `json:"signature,omitempty"`
+}
+
+// CreateSessionToken creates a token for the proxy.
+func CreateSessionToken(username string) (*SessionToken, error) {
+	now := getNow()
+	token := &SessionToken{
+		Username:   username,
+		SignTime:   now,
+		ExpireTime: now.Add(tokenLifetime),
+	}
+	tokenBytes, err := json.Marshal(token)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if token.Signature, err = globalSigningCert.sign(tokenBytes); err != nil {
+		return nil, ErrCannotMigrateSession.GenWithStackByArgs(err.Error())
+	}
+	return token, nil
+}
+
+// ValidateSessionToken validates the token sent from the proxy.
+func ValidateSessionToken(tokenBytes []byte, username string) (err error) {
+	var token SessionToken
+	if err = json.Unmarshal(tokenBytes, &token); err != nil {
+		return errors.Trace(err)
+	}
+	signature := token.Signature
+	// Clear the signature and marshal it again to get the original content.
+	token.Signature = nil
+	if tokenBytes, err = json.Marshal(token); err != nil {
+		return errors.Trace(err)
+	}
+	if err = globalSigningCert.checkSignature(tokenBytes, signature); err != nil {
+		return ErrCannotMigrateSession.GenWithStackByArgs(err.Error())
+	}
+	now := getNow()
+	if now.After(token.ExpireTime) {
+		return ErrCannotMigrateSession.GenWithStackByArgs("token expired", token.ExpireTime.String())
+	}
+	// This prevents the creation of a token by brute-force.
+	if token.SignTime.Add(tokenLifetime).Before(now) {
+		return ErrCannotMigrateSession.GenWithStackByArgs("token lifetime is too long", token.SignTime.String())
+	}
+	if !strings.EqualFold(username, token.Username) {
+		return ErrCannotMigrateSession.GenWithStackByArgs("username does not match", username, token.Username)
+	}
+	return nil
+}
+
+// SetKeyPath sets the path of key.pem and force load the certificate again.
+func SetKeyPath(keyPath string) {
+	globalSigningCert.setKeyPath(keyPath)
+}
+
+// SetCertPath sets the path of key.pem and force load the certificate again.
+func SetCertPath(certPath string) {
+	globalSigningCert.setCertPath(certPath)
+}
+
+// StartCertLoader starts a goroutine to load the certificate periodically.
+// It's impossible to know when the old certificate should expire without this goroutine:
+// - If the certificate is rotated a minute ago, the old certificate should be still valid for a while.
+// - If the certificate is rotated a month ago, the old certificate should expire for safety.
+func StartCertLoader() {
+	go func() {
+		for range time.Tick(loadCertInterval) {
+			globalSigningCert.checkAndLoadCert()
+		}
+	}()
+}
```

Comment:
Maybe we should add `StopCertLoader` to prevent go routine leaking. But that can be done in the next PR.

Reply:
I replaced the function. I'll make the loop in `domain` in a future PR so that I can use the channel `domain.exit` to stop the loader.